Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift

```diff
@@ -1257,100 +1321,133 @@ private final class PubGrubPackageContainer {
             return result
         }
 
-        let versions: [Version] = try packageContainer.reversedVersions().reversed()
+        let versions: [Version] = try self.underlying.reversedVersions().reversed()
 
         // This is guaranteed to be present.
         let idx = versions.firstIndex(of: fromVersion)!
 
         // Compute upper and lower bounds for the dependencies.
-        let upperBounds = computeBounds(with: AnyCollection(versions.dropFirst(idx + 1)), upperBound: true)
-        let lowerBounds = computeBounds(with: AnyCollection(versions.dropLast(versions.count - idx).reversed()), upperBound: false)
 
-        return (lowerBounds, upperBounds)
+        let sync = DispatchGroup()
+
+        var upperBounds: [PackageReference: Version] = [:]
+        self.queue.async(group: sync) {
+            upperBounds = compute(AnyCollection(versions.dropFirst(idx + 1)), upperBound: true)
+        }
+
+        var lowerBounds: [PackageReference: Version] = [:]
+        self.queue.async(group: sync) {
+            lowerBounds = compute(AnyCollection(versions.dropLast(versions.count - idx).reversed()), upperBound: false)
+        }
+
+        // FIXME: should this timeout be configurable?
+        switch sync.wait(timeout: .now() + 60) {
```

Comment:
It probably should.  Also, would there be any callbacks during this time?  If this is waiting for fetches to complete then those would be enough, but if there might be up to a minute of no output, we should make sure to show what is going on (perhaps some intermediate timeout and then a message, followed by a longer timeout).

Reply:
computeBounds is our bottle neck: it iterates through the available tags/versions and loads the manifest to check for things like tools version compatibility etc. the timeout is on that operation, its not network bound but I/o bound.

Reply:
Ok, I see.  That makes sense.  So the only possibility of that taking arbitrarily long is either I/O or some internal logic error causing a loop.

Reply:
yes, disk operations or thread starvation